### DISCUSSION TITLE
How to add Lod Mesh？

### DISCUSSION THREAD

**Discussion by flyskyosg** [o3de/o3de#18100] [Help Wanted]
I have some vertices, index data, how to add to the scene, and support instantiation rendering?

**Comment by idclosed** [comment]
void ModelAssetHelpers::CreateModel(
    ModelAsset* modelAsset,
    const AZ::Name& name,
    AZStd::span<const uint32_t> indices,
    AZStd::span<const float_t> positions,
    AZStd::span<const float> normals,
    AZStd::span<const float> tangents,
    AZStd::span<const float> bitangents,
    AZStd::span<const float> uvs)
{
    // First build a model LOD asset that contains a mesh for the given data.
    AZ::Data::Asset<AZ::RPI::ModelLodAsset> lodAsset;

    AZ::Data::AssetId lodAssetId = AZ::Uuid::CreateRandom();
    lodAsset = AZ::Data::AssetManager::Instance().CreateAsset(
        lodAssetId, azrtti_typeid<AZ::RPI::ModelLodAsset>(), Data::AssetLoadBehavior::PreLoad);

    AZ::RPI::ModelLodAssetCreator creator;
    creator.Begin(lodAssetId);

    const uint32_t positionElementCount = aznumeric_cast<uint32_t>(positions.size() / 3);
    const uint32_t indexElementCount = aznumeric_cast<uint32_t>(indices.size());
    const uint32_t uvElementCount = aznumeric_cast<uint32_t>(uvs.size() / 2);
    const uint32_t normalElementCount = aznumeric_cast<uint32_t>(normals.size() / 3);
    const uint32_t tangentElementCount = aznumeric_cast<uint32_t>(tangents.size() / 4);
    const uint32_t bitangentElementCount = aznumeric_cast<uint32_t>(bitangents.size() / 3);

    constexpr uint32_t PositionElementSize = aznumeric_cast<uint32_t>(sizeof(float) * 3);
    constexpr uint32_t IndexElementSize = aznumeric_cast<uint32_t>(sizeof(uint32_t));
    constexpr uint32_t UvElementSize = aznumeric_cast<uint32_t>(sizeof(float) * 2);
    constexpr uint32_t NormalElementSize = aznumeric_cast<uint32_t>(sizeof(float) * 3);
    constexpr uint32_t TangentElementSize = aznumeric_cast<uint32_t>(sizeof(float) * 4);
    constexpr uint32_t BitangentElementSize = aznumeric_cast<uint32_t>(sizeof(float) * 3);

    // Calculate the Aabb for the given positions.
    AZ::Aabb aabb = AZ::Aabb::CreateNull();
    for (uint32_t i = 0; i < positions.size(); i += 3)
    {
        aabb.AddPoint(AZ::Vector3(positions[i], positions[i + 1], positions[i + 2]));
    }

    // Set up a single-mesh asset with only position data.
    creator.BeginMesh();
    creator.SetMeshAabb(AZStd::move(aabb));
    creator.SetMeshMaterialSlot(0);
    creator.SetMeshIndexBuffer({ CreateBufferAsset(indices.data(), indexElementCount, IndexElementSize),
                                 AZ::RHI::BufferViewDescriptor::CreateTyped(0, indexElementCount, AZ::RHI::Format::R32_UINT) });
    creator.AddMeshStreamBuffer(
        AZ::RHI::ShaderSemantic(AZ::Name("POSITION")),
        AZ::Name(),
        { CreateBufferAsset(positions.data(), positionElementCount, PositionElementSize),
          AZ::RHI::BufferViewDescriptor::CreateTyped(0, positionElementCount, AZ::RHI::Format::R32G32B32_FLOAT) });
    creator.AddMeshStreamBuffer(
        AZ::RHI::ShaderSemantic(AZ::Name("NORMAL")),
        AZ::Name(),
        { CreateBufferAsset(normals.data(), normalElementCount, NormalElementSize),
          AZ::RHI::BufferViewDescriptor::CreateTyped(0, normalElementCount, AZ::RHI::Format::R32G32B32_FLOAT) });
    creator.AddMeshStreamBuffer(
        AZ::RHI::ShaderSemantic(AZ::Name("TANGENT")),
        AZ::Name(),
        { CreateBufferAsset(tangents.data(), tangentElementCount, TangentElementSize),
          AZ::RHI::BufferViewDescriptor::CreateTyped(0, tangentElementCount, AZ::RHI::Format::R32G32B32A32_FLOAT) });
    creator.AddMeshStreamBuffer(
        AZ::RHI::ShaderSemantic(AZ::Name("BITANGENT")),
        AZ::Name(),
        { CreateBufferAsset(bitangents.data(), bitangentElementCount, BitangentElementSize),
          AZ::RHI::BufferViewDescriptor::CreateTyped(0, bitangentElementCount, AZ::RHI::Format::R32G32B32_FLOAT) });
    creator.AddMeshStreamBuffer(
        AZ::RHI::ShaderSemantic(AZ::Name("UV")),
        AZ::Name(),
        { CreateBufferAsset(uvs.data(), uvElementCount, UvElementSize),
          AZ::RHI::BufferViewDescriptor::CreateTyped(0, uvElementCount, AZ::RHI::Format::R32G32_FLOAT) });
    creator.EndMesh();
    creator.End(lodAsset);

    // Create a model asset that contains the single LOD built above.
    modelAsset->InitData(
        name,
        AZStd::span<AZ::Data::Asset<AZ::RPI::ModelLodAsset>>(&lodAsset, 1),
        {}, // no material slots
        {}, // no fallback material
        {} // no tags
    );
}
